### ISSUE TITLE
Undefined Symbol nlopt_set_maxtime

### ISSUE THREAD

**Issue by myloginid** [astamm/nloptr#15]
Hello,

I get the below error while installing nloptr_1.0.4.tar.gz using R CMD INSTALL...

*\* testing if installed package can be loaded
Error in dyn.load(file, DLLpath = DLLpath, ...) :
  unable to load shared object '/usr/lib64/RRO-8.0.2/R-3.1.2/lib64/R/library/nloptr/libs/nloptr.so':
  /usr/lib64/RRO-8.0.2/R-3.1.2/lib64/R/library/nloptr/libs/nloptr.so: undefined symbol: nlopt_set_maxtime
Error: loading failed
Execution halted
ERROR: loading failed
- removing ‘/usr/lib64/RRO-8.0.2/R-3.1.2/lib64/R/library/nloptr’
  [mm@x01tbipapp3a ~]$

Any clues?

thanks,
Manish

**Comment by jyypma** [comment]
Hi Manish,

Could you provide a bit more information, such as the complete log of the installation, the operating system you are on, and the method that you used to install NLopt?

Thanks,
Jelmet

-----Oorspronkelijk bericht-----
Van: "Manish Maheshwari" notifications@github.com
Verzonden: ‎20-‎10-‎2015 06:44
Aan: "jyypma/nloptr" nloptr@noreply.github.com
Onderwerp: [nloptr] Undefined Symbol nlopt_set_maxtime (#15)

Hello,
I get the below error while installing nloptr_1.0.4.tar.gz using R CMD INSTALL...
*\* testing if installed package can be loaded
Error in dyn.load(file, DLLpath = DLLpath, ...) :
unable to load shared object '/usr/lib64/RRO-8.0.2/R-3.1.2/lib64/R/library/nloptr/libs/nloptr.so':
/usr/lib64/RRO-8.0.2/R-3.1.2/lib64/R/library/nloptr/libs/nloptr.so: undefined symbol: nlopt_set_maxtime
Error: loading failed
Execution halted
ERROR: loading failed
removing ‘/usr/lib64/RRO-8.0.2/R-3.1.2/lib64/R/library/nloptr’ [mm@x01tbipapp3a ~]$ 
Any clues?
thanks,
Manish

**Comment by karmarv** [comment]
Hi @jyypma 

I fell for the same issue. Even upon trying the steps by @FlorianSchwendinger it did not work.

I have already installed*\* nlopt-2.4.2 **in /usr/local/lib using
./configure --enable-shared
make 
sudo make install

Verified the /usr/local/lib to have the following objects upon installation. This is also added to LD_LIBRARY_PATH
ls /usr/local/lib/libn*
libnlopt.a         libnlopt.la        libnlopt.so        libnlopt.so.0      libnlopt.so.0.8.2

I have attempted to install twice. One is installation from source and other is from github using the devtools package.

**1.)  Here is the first error case when i attempt to install nloptr_1.0.4.tar.gz from source**
`

```
[dbadmin@vertica-team devtools]$ sudo /opt/vertica/R/bin/R CMD INSTALL nloptr/
* installing to library ‘/opt/vertica/R/library’
* installing *source* package ‘nloptr’ ...
file ‘src/Makevars.in’ has the wrong MD5 checksum
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking how to run the C++ preprocessor... g++ -E
checking whether we are using the GNU C++ compiler... (cached) yes
checking whether g++ accepts -g... (cached) yes
checking for pkg-config... yes
configure: Now testing for NLopt header file.
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking nlopt.h usability... yes
checking nlopt.h presence... yes
checking for nlopt.h... yes
configure: Suitable NLopt library found.
configure: creating ./config.status
config.status: creating src/Makevars
** libs
make: Nothing to be done for `all'.
installing to /opt/vertica/R/library/nloptr/libs
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Error in dyn.load(file, DLLpath = DLLpath, ...) :
  unable to load shared object '/opt/vertica/R/library/nloptr/libs/nloptr.so':
  /opt/vertica/R/library/nloptr/libs/nloptr.so: undefined symbol: nlopt_set_maxtime
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/opt/vertica/R/library/nloptr’
```

`

**2.) In my case I am lucky to have an internet connection so that I can test out the installation from the git repo. Here is the log when I try to use install_github with the devtools library.**

`

```
> library('devtools')
> install_github("jyypma/nloptr")
Downloading GitHub repo jyypma/nloptr@master
from URL https://api.github.com/repos/jyypma/nloptr/zipball/master
Installing nloptr
'/opt/vertica/R/bin/R' --no-site-file --no-environ --no-save --no-restore CMD  \
  INSTALL '/tmp/RtmpX15lex/devtools1ba36d5286a/jyypma-nloptr-dc6c3ba'  \
  --library='/opt/vertica/R/library' --install-tests

* installing *source* package ‘nloptr’ ...
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking how to run the C++ preprocessor... g++ -E
checking whether we are using the GNU C++ compiler... (cached) yes
checking whether g++ accepts -g... (cached) yes
checking for pkg-config... yes
configure: Now testing for NLopt header file.
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking nlopt.h usability... yes
checking nlopt.h presence... yes
checking for nlopt.h... yes
configure: Now testing for NLopt versison number.
configure: Suitable NLopt library found.
configure: creating ./config.status
config.status: creating src/Makevars
config.status: creating R/PkgFlags.R
** libs
g++ -I/opt/vertica/R/include -DNDEBUG  -I/usr/local/include    -fpic  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -D_FORTIFY_SOURCE=2 -g  -c dummy.cpp -o dummy.o
gcc -std=gnu99 -I/opt/vertica/R/include -DNDEBUG  -I/usr/local/include    -fpic  -g -O2  -c nloptr.c -o nloptr.o
g++ -shared -L/opt/vertica/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o nloptr.so dummy.o nloptr.o -L/opt/vertica/R/lib -lR
installing to /opt/vertica/R/library/nloptr/libs
** R
** inst
** tests
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Error in dyn.load(file, DLLpath = DLLpath, ...) :
  unable to load shared object '/opt/vertica/R/library/nloptr/libs/nloptr.so':
  /opt/vertica/R/library/nloptr/libs/nloptr.so: undefined symbol: nlopt_set_maxtime
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/opt/vertica/R/library/nloptr’
Error: Command failed (1)
```

`

It just doesnt load the shared object in any case. 
Let me know if there is something else that I may try here.

**Comment by jyypma** [comment]
Could you provide the output of

pkg-config --exists nlopt
pkg-config --libs nlopt
pkg-config --cflags nlopt
pkg-config --modversion nlopt

Thanks.

On Tue, Mar 1, 2016 at 11:06 AM, Rahul Vishwakarma <notifications@github.com

**Comment by karmarv** [comment]
Hi,

Here is the output...

```
[dbadmin@vertica-team root]$ pkg-config --exists nlopt
[dbadmin@vertica-team root]$ pkg-config --libs nlopt
Package nlopt was not found in the pkg-config search path.
Perhaps you should add the directory containing `nlopt.pc'
to the PKG_CONFIG_PATH environment variable
No package 'nlopt' found
[dbadmin@vertica-team root]$ pkg-config --cflags nlopt
Package nlopt was not found in the pkg-config search path.
Perhaps you should add the directory containing `nlopt.pc'
to the PKG_CONFIG_PATH environment variable
No package 'nlopt' found
[dbadmin@vertica-team root]$ pkg-config --modversion nlopt
Package nlopt was not found in the pkg-config search path.
Perhaps you should add the directory containing `nlopt.pc'
to the PKG_CONFIG_PATH environment variable
No package 'nlopt' found
```

Considering the hint, I added the environment variables as export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig/ and re ran the  installation only to face the same error.

```

** testing if installed package can be loaded
Error in dyn.load(file, DLLpath = DLLpath, ...) :
  unable to load shared object '/opt/vertica/R/library/nloptr/libs/nloptr.so':
  /opt/vertica/R/library/nloptr/libs/nloptr.so: undefined symbol: nlopt_set_maxtime
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/opt/vertica/R/library/nloptr’
[dbadmin@vertica-team rpackages]$ echo $PKG_CONFIG_PATH
/usr/local/lib/pkgconfig/

```

Here is the pkg-config state after updating the environment variable,

```
[dbadmin@vertica-team rpackages]$ pkg-config --exists nlopt
[dbadmin@vertica-team rpackages]$ pkg-config --libs nlopt
-L/usr/local/lib -lnlopt -lm
[dbadmin@vertica-team rpackages]$ pkg-config --cflags nlopt
-I/usr/local/include
[dbadmin@vertica-team rpackages]$ pkg-config --modversion nlopt
2.4.2
```

**Comment by jyypma** [comment]
Could you show the complete log of the installation process?

**Comment by pacosoplas** [comment]
Hi:
I have the same problem, i did everything , here is the log

[root@lnxbig05 ROBERTO]# export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig/^C
[root@lnxbig05 ROBERTO]# pkg-config --exists nlopt
[root@lnxbig05 ROBERTO]# pkg-config --libs nlopt
-L/usr/local/lib -lnlopt -lm
[root@lnxbig05 ROBERTO]#  pkg-config --cflags nlopt
-I/usr/local/include
[root@lnxbig05 ROBERTO]# R CMD INSTALL nloptr
- installing to library ‘/usr/lib64/R/library’
- installing _source_ package ‘nloptr’ ...
  file ‘configure’ has the wrong MD5 checksum
  checking for g++... g++
  checking whether the C++ compiler works... yes
  checking for C++ compiler default output file name... a.out
  checking for suffix of executables...
  checking whether we are cross compiling... no
  checking for suffix of object files... o
  checking whether we are using the GNU C++ compiler... yes
  checking whether g++ accepts -g... yes
  checking how to run the C++ preprocessor... g++ -E
  checking whether we are using the GNU C++ compiler... (cached) yes
  checking whether g++ accepts -g... (cached) yes
  checking for pkg-config... yes
  configure: Now testing for NLopt header file.
  checking for grep that handles long lines and -e... /bin/grep
  checking for egrep... /bin/grep -E
  checking for ANSI C header files... yes
  checking for sys/types.h... yes
  checking for sys/stat.h... yes
  checking for stdlib.h... yes
  checking for string.h... yes
  checking for memory.h... yes
  checking for strings.h... yes
  checking for inttypes.h... yes
  checking for stdint.h... yes
  checking for unistd.h... yes
  checking nlopt.h usability... yes
  checking nlopt.h presence... yes
  checking for nlopt.h... yes
  configure: Suitable NLopt library found.
  configure: creating ./config.status
  config.status: creating src/Makevars
  *\* libs
  make: Nothing to be done for `all'.
  installing to /usr/lib64/R/library/nloptr/libs
  *\* R
  *\* inst
  *\* preparing package for lazy loading
  *\* help
  **\* installing help indices
  converting help for package ‘nloptr’
    finding HTML links ... done
    auglag                                  html
    bobyqa                                  html
    check.derivatives                       html
    cobyla                                  html
    crs2lm                                  html
    direct                                  html
    gradients                               html
    is.nloptr                               html
    isres                                   html
    lbfgs                                   html
    mlsl                                    html
    mma                                     html
    neldermead                              html
    newuoa                                  html
    nloptions                               html
    nloptr-package                          html
    nloptr                                  html
    nloptr.get.default.options              html
    nloptr.print.options                    html
    print.nloptr                            html
    sbplx                                   html
    slsqp                                   html
    stogo                                   html
    tnewton                                 html
    varmetric                               html
  *\* building package indices
  *\* installing vignettes
  *\* testing if installed package can be loaded
  Error in dyn.load(file, DLLpath = DLLpath, ...) :
  unable to load shared object '/usr/lib64/R/library/nloptr/libs/nloptr.so':
  /usr/lib64/R/library/nloptr/libs/nloptr.so: undefined symbol: nlopt_set_maxtime
  Error: loading failed
  Execution halted
  ERROR: loading failed
- removing ‘/usr/lib64/R/library/nloptr’

**Comment by jyypma** [comment]
This error usually occurs when NLopt is installed in a non-default location
(i.e. not in LD_LIBRARY_PATH). Can you show the output of
  ldconfig -p | grep nlopt

Another option may be that a previously installed version of NLopt is used
during installation. Could you check which version of NLopt is installed?
pkg-config --modversion nlopt

The R interface only works with version 2.4.0 or higher. The version of the
interface that is on CRAN performs this check incorrectly during
configuration.

Finally, it sometimes helps to `make uninstall` NLopt to remove any old
versions that are installed on the system.

**Comment by chansonZ** [comment]
# when i export,then **success**,lib_path: the path of libnlopt.so
export LD_LIBRARY_PATH="lib_path":${LD_LIBRARY_PATH}
export PKG_CONFIG_PATH= "lib_path"/pkgconfig/

**Comment by taranu** [comment]
I had the same issue on a server using modules without having one set up for nlopt. Adding /usr/lib64 to LD_LIBRARY_PATH and /usr/lib64/pkgconfig to PKG_CONFIG_PATH like chansonZ suggested solved the problem.

**Comment by lambdamoses** [comment]
I have a very similar problem. I first tried `install.packages("nloptr")`, and got an error that nlopt.h is not found. Then I googled for a solution, and installed NLopt 2.4.2 with 
```
./configure
make
sudo make install
```
in the default location, /usr/local/lib, according to an answer I found and no longer got that error. Then instead, I got this error:
```
Error: package or namespace load failed for ‘nloptr’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/Users/lambda/anaconda3/lib/R/library/nloptr/libs/nloptr.so':
  dlopen(/Users/lambda/anaconda3/lib/R/library/nloptr/libs/nloptr.so, 6): Symbol not found: _nlopt_add_equality_mconstraint
  Referenced from: /Users/lambda/anaconda3/lib/R/library/nloptr/libs/nloptr.so
  Expected in: flat namespace
 in /Users/lambda/anaconda3/lib/R/library/nloptr/libs/nloptr.so
Error: loading failed
Execution halted
```
Then I used `Sys.setenv` to add /usr/local/lib to LD_LIBRARY_PATH and /usr/local/lib/pkgconfig to PKG_CONFIG_PATH, and tried to install nloptr again using install.packages, I still got the same error. 
The full message I got when installing is:
```
> install.packages('nloptr', dependencies = T)
trying URL 'https://cran.rstudio.com/src/contrib/nloptr_1.0.4.tar.gz'
Content type 'application/x-gzip' length 353957 bytes (345 KB)
==================================================
downloaded 345 KB

* installing *source* package ‘nloptr’ ...
** package ‘nloptr’ successfully unpacked and MD5 sums checked
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking how to run the C++ preprocessor... g++ -E
checking whether we are using the GNU C++ compiler... (cached) yes
checking whether g++ accepts -g... (cached) yes
checking for pkg-config... no
configure: Now testing for NLopt header file.
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking nlopt.h usability... yes
checking nlopt.h presence... yes
checking for nlopt.h... yes
configure: Suitable NLopt library found.
configure: creating ./config.status
config.status: creating src/Makevars
** libs
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -I/Users/lambda/anaconda3/lib/R/include -DNDEBUG   -D_FORTIFY_SOURCE=2 -mmacosx-version-min=10.9 -I/Users/lambda/anaconda3/include   -fPIC  -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden  -fmessage-length=0 -I/Users/lambda/anaconda3/include  -c dummy.cpp -o dummy.o
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -I/Users/lambda/anaconda3/lib/R/include -DNDEBUG   -D_FORTIFY_SOURCE=2 -mmacosx-version-min=10.9 -I/Users/lambda/anaconda3/include   -fPIC  -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -I/Users/lambda/anaconda3/include  -c nloptr.c -o nloptr.o
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/Users/lambda/anaconda3/lib/R/lib -Wl,-pie -Wl,-headerpad_max_install_names -o nloptr.so dummy.o nloptr.o -L/Users/lambda/anaconda3/lib/R/lib -lR -L/Users/lambda/anaconda3/lib -lintl -lc -Wl,-framework -Wl,CoreFoundation
ld: warning: -pie being ignored. It is only used when linking a main executable
installing to /Users/lambda/anaconda3/lib/R/library/nloptr/libs
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Error: package or namespace load failed for ‘nloptr’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/Users/lambda/anaconda3/lib/R/library/nloptr/libs/nloptr.so':
  dlopen(/Users/lambda/anaconda3/lib/R/library/nloptr/libs/nloptr.so, 6): Symbol not found: _nlopt_add_equality_mconstraint
  Referenced from: /Users/lambda/anaconda3/lib/R/library/nloptr/libs/nloptr.so
  Expected in: flat namespace
 in /Users/lambda/anaconda3/lib/R/library/nloptr/libs/nloptr.so
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/Users/lambda/anaconda3/lib/R/library/nloptr’
Warning in install.packages :
  installation of package ‘nloptr’ had non-zero exit status

The downloaded source packages are in
	‘/private/var/folders/tg/cv0dwcv11q5g13yjt9mc_nh00000gn/T/RtmpKOvmWV/downloaded_packages’
Updating HTML index of packages in '.Library'
Making 'packages.html' ... done
```
These are the outputs from pkg-config:
```
Lambdas-MacBook-Pro:lib lambda$ pkg-config --exists nlopt
Lambdas-MacBook-Pro:lib lambda$ pkg-config --libs nlopt
-L/usr/local/lib -lnlopt -lm
Lambdas-MacBook-Pro:lib lambda$ pkg-config --cflags nlopt
-I/usr/local/include
Lambdas-MacBook-Pro:lib lambda$ pkg-config --modversion nlopt
2.4.2
```
In addition, when installing NLopt, I got a lot of messages that functions are deprecated, such as 
```
In file included from f77api.c:103:
./f77funcs.h:57:14: warning: 'nlopt_minimize_constrained' is deprecated
  [-Wdeprecated-declarations]
 *info = nlopt_minimize_constrained((nlopt_algorithm) *algorithm, 
         ^
./nlopt.h:348:35: note: 'nlopt_minimize_constrained' has been explicitly marked
  deprecated here
 int maxeval, double maxtime) NLOPT_DEPRECATED;
                              ^
./nlopt.h:320:43: note: expanded from macro 'NLOPT_DEPRECATED'
#  define NLOPT_DEPRECATED __attribute__((deprecated))`
                                      ^
```
I read in another issue here that `sudo apt-get install libnlopt-dev` fixed the issue on Ubuntu, but there's no apt-get on Mac and I couldn't get lib nlopt-dev from homebrew.
Can someone please help?

**Comment by jorgeboucas** [comment]
We are also facing the same problem on centos 7.3.1611

```
** testing if installed package can be loaded
Error in dyn.load(file, DLLpath = DLLpath, ...) : 
  unable to load shared object '/beegfs/group_bit/home/JBoucas/.R/3.3.2/R_LIBS_USER/nloptr/libs/nloptr.so':
  /beegfs/group_bit/home/JBoucas/.R/3.3.2/R_LIBS_USER/nloptr/libs/nloptr.so: undefined symbol: nlopt_remove_equality_constraints
```

Here the env.:
```
amaliax:~$ echo $LD_LIBRARY_PATH
/beegfs/group_bit/home/JBoucas/nlopt_test/lib64
amaliax:~$ echo $CPATH
/beegfs/group_bit/home/JBoucas/nlopt_test/include
amaliax:~$ echo $CFLAGS
-I/beegfs/group_bit/home/JBoucas/nlopt_test/include
amaliax:~$ echo $LDFLAGS
-L/beegfs/group_bit/home/JBoucas/nlopt_test/lib64 
```

**Comment by lambdamoses** [comment]
Update: the problem went away as I reinstalled Anaconda, and installed R and RStudio outside Anaconda. It seems that Anaconda or the RStudio downloaded from the condo repo is causing the problem in my case. I also had problem installing RStan, but it went away as well after this.

**Comment by astamm** [comment]
We just released `v2.0.0`. We now rely on `Cmake` when building from source. Can you test and tell us if it fixes whatever issue you had in the past please?
